File: src/main/java/fr/raconteur/simpleskinswapper/SimpleSkinSwapper.java
package fr.raconteur.simpleskinswapper;

import fr.raconteur.simpleskinswapper.changeskin.IPlayerSkinUpdatable;
import fr.raconteur.simpleskinswapper.networking.HandshakePayload;
import fr.raconteur.simpleskinswapper.networking.SkinRefreshPayload;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry;
import net.fabricmc.fabric.api.networking.v1.ServerPlayConnectionEvents;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.network.packet.s2c.play.PlayerListS2CPacket;
import net.minecraft.network.packet.s2c.play.PlayerRemoveS2CPacket;
import net.minecraft.server.network.ServerPlayerEntity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collections;

public class SimpleSkinSwapper implements ModInitializer {

    public static final String MOD_ID = "simpleskinswapper";
    public static final Logger LOGGER = LoggerFactory.getLogger(MOD_ID);

    @Override
    public void onInitialize() {
        PayloadTypeRegistry.playC2S().register(SkinRefreshPayload.PACKET_ID, SkinRefreshPayload.PACKET_CODEC);
        PayloadTypeRegistry.playS2C().register(HandshakePayload.PACKET_ID, PacketCodec.unit(HandshakePayload.INSTANCE));

        // Announce mod presence to client (enables direct packet path in singleplayer and Fabric servers)
        ServerPlayConnectionEvents.JOIN.register((handler, sender, server) -> {
            ServerPlayerEntity player = handler.player;
            if (ServerPlayNetworking.canSend(player, HandshakePayload.PACKET_ID)) {
                ServerPlayNetworking.send(player, HandshakePayload.INSTANCE);
            }
        });

        // Handle skin refresh packets sent by the client
        ServerPlayNetworking.registerGlobalReceiver(SkinRefreshPayload.PACKET_ID, (payload, context) -> {
            ServerPlayerEntity player = context.player();
            LOGGER.info("Received SkinRefreshPayload from {}", player.getName().getString());
            context.server().execute(() -> {
                ((IPlayerSkinUpdatable) player).simpleSkinSwapper$setGameProfileWithTexture(payload.textureProperty());

                var playerManager = context.server().getPlayerManager();
                playerManager.sendToAll(new PlayerRemoveS2CPacket(new ArrayList<>(Collections.singleton(player.getUuid()))));
                playerManager.sendToAll(PlayerListS2CPacket.entryFromPlayer(Collections.singleton(player)));
            });
        });

        LOGGER.info("SimpleSkinSwapper initializing...");
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/SimpleSkinSwapper.kt
package fr.raconteur.simpleskinswapper

import fr.raconteur.simpleskinswapper.changeskin.IPlayerSkinUpdatable
import fr.raconteur.simpleskinswapper.networking.HandshakePayload
import fr.raconteur.simpleskinswapper.networking.SkinRefreshPayload
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry
import net.fabricmc.fabric.api.networking.v1.ServerPlayConnectionEvents
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.game.ClientboundPlayerInfoRemovePacket
import net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class SimpleSkinSwapper : ModInitializer {

    override fun onInitialize() {
        PayloadTypeRegistry.serverboundPlay().register(SkinRefreshPayload.PACKET_ID, SkinRefreshPayload.PACKET_CODEC)
        PayloadTypeRegistry.clientboundPlay().register(HandshakePayload.PACKET_ID, StreamCodec.unit(HandshakePayload))

        // Announce mod presence to client (enables direct packet path in singleplayer and Fabric servers)
        ServerPlayConnectionEvents.JOIN.register { handler, _, _ ->
            val player = handler.player
            if (ServerPlayNetworking.canSend(player, HandshakePayload.PACKET_ID)) {
                ServerPlayNetworking.send(player, HandshakePayload)
            }
        }

        // Handle skin refresh packets sent by the client
        ServerPlayNetworking.registerGlobalReceiver(SkinRefreshPayload.PACKET_ID) { payload, context ->
            val player = context.player()
            LOGGER.info("Received SkinRefreshPayload from {}", player.name.string)
            context.server().execute {
                // Replace the GameProfile with a new one carrying the updated texture property.
                // GameProfile is a record in authlib 7+ so its PropertyMap may be immutable;
                // the mixin creates a fresh mutable PropertyMap instead of mutating the existing one.
                (player as IPlayerSkinUpdatable).`simpleSkinSwapper$setGameProfileWithTexture`(payload.textureProperty)

                // Broadcast updated tab list so all clients receive the new GameProfile with texture
                val playerList = context.server().playerList
                playerList.broadcastAll(ClientboundPlayerInfoRemovePacket(listOf(player.uuid)))
                playerList.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(listOf(player)))
            }
        }

        LOGGER.info("SimpleSkinSwapper initializing...")
    }

    companion object {
        const val MOD_ID = "simpleskinswapper"

        @JvmField
        val LOGGER: Logger = LoggerFactory.getLogger(MOD_ID)
    }
}
